File: Data Structures & Algorithms/pow-x-n/submission-1.java
class Solution {
    public double myPow(double x, int n) {
       if(x == 0 && n > 0){
         return 0;
       }

       if(n < 0){
         return 1/helper(x,-1*n);
       }

       return helper(x,n);
    }

    public double helper(double x, int n) {
        if(n == 0){
            return 1;
        }
        if(n == 1){
            return x;
        }

        double res = helper(x, n-1);
        return res*x;
    }
}

File: Data Structures & Algorithms/generate-parentheses/submission-1.kt
class Solution {
    var ans = mutableListOf<String>()
    fun generateParenthesis(n: Int): List<String> {
        val sb = StringBuilder()
        helper(n, 0, sb, 0)

        return ans
    }

    fun helper(n: Int, idx: Int, result: StringBuilder, sum: Int) {
        if(sum < 0 || idx > n*2){
            return
        }

        if(idx == n*2 && sum == 0){
            ans.add(result.toString())
            return
        }

        result.append('(')
        helper(n, idx+1, result, sum+1)
        result.setLength(result.length - 1)
        result.append(')')
        helper(n, idx+1, result, sum-1)
        result.setLength(result.length - 1)

        return
    }
}
